esp/esdllib/esdl_script.cpp

```diff
@@ -1671,6 +1671,78 @@ class CEsdlTransformOperationTrace : public CEsdlTransformOperationWithoutChildr
         }
         return false;
     }
+
+    virtual bool getText(StringBuffer& text, IXpathContext* targetContext, IXpathContext* sourceContext, IDataMaskingProfileContext* masker) = 0;
+};
+
+class CEsdlTransformOperationTraceContent : public CEsdlTransformOperationTrace
+{
+protected:
+    StringAttr m_contentType;
+    Owned<ICompiledXpath> m_skipMask;
+
+public:
+    CEsdlTransformOperationTraceContent(IXmlPullParser &xpp, StartTag &stag, const StringBuffer &prefix) : CEsdlTransformOperationTrace(xpp, stag, prefix)
+    {
+        m_contentType.set(stag.getValue("content_type"));
+        const char* skipMask = stag.getValue("skip_mask");
+        if (!isEmptyString(skipMask))
+            m_skipMask.setown(compileXpath(skipMask));
+    }
+
+    virtual bool getText(StringBuffer& text, IXpathContext* targetContext, IXpathContext* sourceContext, IDataMaskingProfileContext* masker) override
+    {
+        bool isValue = false;
+        if (!targetContext->selectText(m_select, text, isValue) || !masker)
+            return true;
+        if (!m_skipMask || !sourceContext->evaluateAsBoolean(m_skipMask))
+        {
+            char* buffer = const_cast<char*>(text.str());
+            masker->maskContent(m_contentType.get(), buffer, 0, text.length());
+        }
+        return true;
+    }
+};
+
+class CEsdlTransformOperationTraceValue : public CEsdlTransformOperationTrace
+{
+protected:
+    XPathLiteralUnion m_valueType;
+    XPathLiteralUnion m_maskStyle;
+
+public:
+    CEsdlTransformOperationTraceValue(IXmlPullParser &xpp, StartTag &stag, const StringBuffer &prefix) : CEsdlTransformOperationTrace(xpp, stag, prefix)
+    {
+        m_valueType.setRequired(stag, "value_type", *this);
+        m_maskStyle.setOptional(stag, "mask_style");
+    }
+
+    virtual bool getText(StringBuffer& text, IXpathContext* targetContext, IXpathContext* sourceContext, IDataMaskingProfileContext* masker) override
+    {
+        bool isValue = true;
+        if (!targetContext->selectText(m_select, text, isValue))
+        {
+            if (!isValue)
+            {
+                recordWarning(ESDL_SCRIPT_Warning, "ambiguous value xpath");
+                return false;
+            }
+        }
+        if (!masker)
```

Comment:
You do some work up above but then just return here.  Should the masker be checked at the top?

Reply:
No. The work above - getting the base text that will be output by the operation - is always required. The absence of a masker is not an error; it means that the base text will be output unchanged. Then we attempt to apply the mask, which could fail.